elf_load.c

```diff
@@ -311,13 +312,132 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
 }
 
 #if CONFIG_LIBVFSCORE
+#if CONFIG_LIBPOSIX_MMAP
+static int elf_load_fdphdr(struct elf_prog *elf_prog, GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart, vaend;
+	uint64_t delta_p_offset;
+	static void *addr;
+	static int flags;
+
+	/* If we are mmapping a fresh ELF, restore address and flags */
+	if (!phdr->p_paddr) {
+		addr = NULL;
+		flags = MAP_PRIVATE;
+	}
+
+	/* PT_LOAD's paddr may be misaligned, so keep in mind the
+	 * offset from what should have been the aligned paddr to
+	 * what we ended up with.
+	 */
+	delta_p_offset = phdr->p_paddr - PAGE_ALIGN_DOWN(phdr->p_paddr);
+
+	/* If this is not the first mmap'd ELF segment, then keep
```

Comment:
If I understand correctly, for the first segment, `addr` will be `NULL`, and the final virtual address will be determined by `mmap`, right? But you didn't tell `mmap` the `p_align` requirement, so it will just guarantee a 4K alignment.

Reply:
Yes, that is correct. But at the moment we only support ELFs with base address 0. :thinking: Do you mean there is an additional `mmap` flag I could've used?

Reply:
> Yes, that is correct. But at the moment we only support ELFs with base address 0. 🤔 Do you mean there is an additional `mmap` flag I could've used?

Should it be `ELFs which are with fPIC (position independent)`?

Maybe we can mmap memory with size `original_size + PAGE_ALIGN_UP(p_align)`. Then there must have a address satisfying the `p_align`, then `munmap` the unused part.

Actually I also can't find a perfect way to alloc these memory without failure. We can just "attempt" to do this, but there's no guarantee for success. Maybe we need to support more flags for `mmap`, in the situation of single address space.

Reply:
Ah, I see, is this the example you had in mind?

```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr           FileSiz            MemSiz         Flags  Align
  LOAD           0x0000000001a00000 0x0000000001a00000 0x0000000001a00000 0x0000000000000231 0x0000000000000231 R E 0x200000
  LOAD           0x0000000001a01000 0x0000000001a02000 0x0000000001a02000 0x0000000000d056f0 0x0000000000d056f0 R   0x200000
  ```
  
  Where basically the start address of the second PT_LOAD minus the end address of the first PT_LOAD is >= PAGE_SIZE? In this case that page that would remain empty between the two would be ignored. See this line
  ```C
  		addr = (void *)PAGE_ALIGN_DOWN((phdr->p_paddr + (uintptr_t)elf_prog->vabase));
```
the start address of the PHDR is taken into account, not the last address we were left at. E.g., for the example above, you would have the following mappings:
```
0x1a00000 -> 0x1a01000
0x1a02000 -> 0x27076f0
```
So the region `0x1a01000 -> 0x1a02000` is left as a hole, neither mapped, nor unmapped, therefore in the `LIBPOSIXMMAP=y` case this resource is not wasted :+1:. However, in the case where you don't have `LIBPOSIXMMAP` and you fallback on reading from the file, that zone does indeed is left mapped and stays there uselessly. However, in such a case, you would not only have to unmap that page, but also `free()` that exact portion of the massive buffer allocated through `memalign` whose size is equal to the size of the file loaded in memory. I don't see how you could easily do this without breaking the memory allocator, since unmapping a random page inside the allocated area might prove to be unsafe depending on the allocator and how it implements `free()`.

> Actually I also can't find a perfect way to alloc these memory without failure. We can just "attempt" to do this, but there's no guarantee for success.

So, I guess, we could just leave this as a drawback of the case where the elfloader uses reading from file instead of `mmap`-ing?

> Maybe we need to support more flags for mmap, in the situation of single address space.

That I do not know much about. Perhaps @skuenzer might have more insight on this. Maybe also worth mentioning during your next app-compat meeting.


What do you think? Also checkout this [comment](https://github.com/unikraft/app-elfloader/pull/28#issuecomment-1742702566) as it might be related your question.

Reply:
> Should it be ELFs which are with fPIC (position independent)?

 To my knowledge, an `fPIC`'s base address is not necessarily `0x0`, but it rather depends on the ABI (e.g. this is a must for ARM bare-metal ABI). For example, you could replace Unikraft's PIE build flags with `PIC` and, since Unikraft's base address on x86 is `0x100000`, you end up with an `fPIC` whose base address is `0x100000` not `0x0`. Actually, this is exactly what Unikraft was 6 months ago before I published the [`PIE` PR](https://github.com/unikraft/unikraft/pull/772): an `fPIC` with base address `0x100000`. Ultimately I pushed for `-static-pie`, because a pure `fPIC` just had waaay too many relocations to handle, unlike a static `PIE` whose only relocations are related to absolute values and GOT, the latter only being for ARM but not relevant in our case.

Reply:
> ```
> Program Headers:
>   Type           Offset             VirtAddr           PhysAddr           FileSiz            MemSiz         Flags  Align
>   LOAD           0x0000000001a00000 0x0000000001a00000 0x0000000001a00000 0x0000000000000231 0x0000000000000231 R E 0x200000
>   LOAD           0x0000000001a01000 0x0000000001a02000 0x0000000001a02000 0x0000000000d056f0 0x0000000000d056f0 R   0x200000
> ```

It's a good example, but I'm not worrying about the "empty hole" between two segments. I'm worrying about the first segment. In this example the `Align` is 0x200000, but `mmap` only guarantees the alignment of 0x1000 (`addr` is NULL at that moment). I think in most cases this doesn't matter, but anyway we are actually breaking the requirements of such ELF files.

Reply:
Or to say, the `elf_prog->vabase` should satisfy the max alignment requirement of all the PT_LOAD segments.

Reply:
> ```
> Program Headers:
>   Type           Offset             VirtAddr           PhysAddr           FileSiz            MemSiz         Flags  Align
>   LOAD           0x0000000001a00000 0x0000000001a00000 0x0000000001a00000 0x0000000000000231 0x0000000000000231 R E 0x200000
>   LOAD           0x0000000001a01000 0x0000000001a02000 0x0000000001a02000 0x0000000000d056f0 0x0000000000d056f0 R   0x200000
> ```
>
> However, in such a case, you would not only have to unmap that page, but also `free()` that exact portion of the massive buffer allocated through `memalign` whose size is equal to the size of the file loaded in memory. I don't see how you could easily do this without breaking the memory allocator, since unmapping a random page inside the allocated area might prove to be unsafe depending on the allocator and how it implements `free()`.

Do you mean the `memalign()` function has some bad side-effects, that make the memory hard to free? We can also avoid using `memalign` by `mmap`ing 2 times of the virtual address space.

For example, we see the requirement of virtual address is:

```
0x0 -> 0x231
0x2000 -> 0x76f0
```
So the total size of continuous virtual address space we need is: `0x76f0`. For alignment `0x200000`, we `mmap` for `0x400000`, and there must be an address aligned to `0x200000` inside the mmap-ed region. Then we `munmap` the unused part, now it is guaranteed that the alignment is satisfied, and there is also no failure from the `mmap`s.

Reply:
Ahh, I see what you mean now. You were referring to the alignment of the first generated `addr`... yes very good catch :+1:. I see that BSD systems do support such a flag for their [`mmap`](https://man.freebsd.org/cgi/man.cgi?mmap(2)) (`MAP_ALIGNED(n)`), but I do not see it in Linux, or its manuals :thinking:. In fact, even before this PR, the call to `memalign` for the initial memory allocation (in the use case of not reading directly from the file).

I guess that what we could do is this:
```
len = maplen >= p_align ? (maplen + p_align) : (p_align << 1);
addr = mmap(NULL, len,...)
addr_aligned = ALIGN_UP(addr, p_align)

// if already aligned just return
if (addr_aligned == addr)
    return addr
    
munmap(addr, len)

addr = mmap(addr_aligned, len, MAP_FIXED)
```

This has the downside of being slow in that rare scenario, but far better than failing. What do you think?
And same goes for the `memalign` scenario it is even easier, we just straight-up allocate enough to include alignment and use the aligned address.

Reply:
We're still inside the elfloader, no need for syscalls, we could use functions from the core :D.

Reply:
Oh, check this out, it may not be so slow after all!

From `man 2 mmap`'s description of `MAP_FIXED`

```
If  the memory region specified by addr and len overlaps pages of any existing mapping(s), then the overlapped  part  of  the  existing  mapping(s) will be discarded.  If the specified address cannot be used, mmap() will fail.
```

And Unikraft seems to replicate this behavior:
<https://github.com/unikraft/unikraft/blob/staging/lib/posix-mmap/mmap.c#L73>
+
<https://github.com/unikraft/unikraft/blob/staging/lib/ukvmem/vmem.c#L639-L645>

So, this should work
```
len = maplen >= p_align ? (maplen + p_align) : (p_align << 1);
addr = mmap(NULL, len,...)
addr_aligned = ALIGN_UP(addr, p_align)

// if already aligned just return
if (addr_aligned == addr)
    return addr
    
munmap(addr, addr_aligned - addr)

return addr_aligned
```

Will provide an update soon.